documentation/src/main/asciidoc/topics/con_jgroups_encrypt.adoc

```diff
@@ -0,0 +1,49 @@
+[id='jgroups_encryption-{context}']
+= {brandname} Cluster Security
+To secure cluster traffic, you configure {brandname} nodes to encrypt JGroups message payloads with secret keys.
+
+{brandname} nodes can obtain secret keys from either:
+
+* The coordinator node (asymmetric encryption).
+* A shared keystore (symmetric encryption).
+
+.Retrieving secret keys from coordinator nodes
+
+You configure asymmetric encryption by adding the `ASYM_ENCRYPT` protocol to a JGroups stack in your {brandname} configuration.
+This allows {brandname} clusters to generate and distribute secret keys.
+
+[IMPORTANT]
+====
+When using asymmetric encryption, you should also provide keystores so that nodes can perform certificate authentication and securely exchange secret keys.
+This protects your cluster from man-in-the-middle (MitM) attacks.
+====
+
+Asymmetric encryption secures cluster traffic as follows:
+
+. The first node in the {brandname} cluster, the coordinator node, generates a secret key.
+. A joining node performs certificate authentication with the coordinator to mutually verify identity.
+. The joining node requests the secret key from the coordinator node. That request includes the public key for the joining node.
+. The coordinator node encrypts the secret key with the public key and returns it to the joining node.
+. The joining node decrypts and installs the secret key.
+. The node joins the cluster, encrypting and decrypting messages with the secret key.
+
+.Retrieving secret keys from shared keystores
+
+You configure symmetric encryption by adding the `SYM_ENCRYPT` protocol to a JGroups stack in your {brandname} configuration.
+This allows {brandname} clusters to obtain secret keys from keystores that you provide.
+
+. Nodes install the secret key from a keystore on the {brandname} classpath at startup.
+. Node join clusters, encrypting and decrypting messages with the secret key.
+
+.Comparison of asymmetric and symmetric encryption
+
+`ASYM_ENCRYPT` with certificate authentication provides an additional layer of encryption in comparison with `SYM_ENCRYPT`.
+You provide keystores that encrypt the requests to coordinator nodes for the secret key.
+{brandname} automatically generates that secret key and handles cluster traffic, while letting you specify when to generate secret keys.
+For example, you can configure clusters to generate new secret keys when nodes leave.
+This ensures that nodes cannot bypass certificate authentication and join with old keys.
+
+`SYM_ENCRYPT`, on the other hand, is faster than `ASYM_ENCRYPT` because nodes install secret keys directly from keystores, rather than over the network.
```

Comment:
@belaban can you comment on what "faster" means here? If I understood the description properly, the asymmetric algorithm is only used to exchange the session key. Then, all messages are encrypted with a symmetric algorithm (so as in `SYM_ENCRYPT`); so the performance should be the same (assuming the same sym algo)

Reply:
Faster means that establishment of a connection is faster with SYM_ENCRYPT, as we don't need to exchange keys first. After that, during regular operation, both protocols have the same speed.

Reply:
Thanks for the comments here, guys. @pruivo Does the "because nodes install ..." phrase not convey the reason why SYM_ENCRYPT is faster? Each node just gets the key from the keystore on it's classpath instead of having to do key exchange with the coordinator, because joining nodes send their public keys to get the secret key from the coordinator. Actually I think I'll just reword to something like this. Cheers. :smile_cat:

Reply:
@oraNod you wanted to be sure if there isn't anything weird happening in the background. 
Usually, the network latency is quite low, so the time to install the key from the coordinator or the latency accessing the disk to install from the file (keystore) shouldn't differ much (IMO, of course, I could be wrong eh eh).

Reply:
@pruivo I got this from Bela who said "I would also document SYM_ENCRYPT, as it is faster than ASYM_ENCRYPT." I've reworded this a little. Maybe I should trim things down some more and say:

"`SYM_ENCRYPT`, on the other hand, is faster than `ASYM_ENCRYPT` because nodes do not need to perform key exchange with the cluster coordinator." 

And drop the bit about "install secret keys directly from keystores on their classpaths" - I see your point about latency. WDYT?

Reply:
sound good :+1: